### PR TITLE
refactor(client): Extract `GroupKeyQueue`

### PR DESCRIPTION
**Pull request by teogeb**
Extracted group key queuing logic to a separate class. 

Before this PR the logic was implemented in `GroupKeyStore`. That class had two separate roles: handling key persistence and managing the group key queue. The extracted `GroupKeyQueue` class uses `GroupKeyStore` to persist the generated class, but is otherwise an independent entity.

Moved all test cases from `GroupKeyStore.test.ts` to a new `GroupKeyQueue.test.ts`, and therefore renamed `GroupKeyQueue2.test.ts` to `GroupKeyQueue.test.ts`.

Other changes:
- `GroupKeyStore#add` returns `Promise<void>`
- inlined `GroupKeyStore#storeKey`

### CODE REVIEW

File: packages/client/src/publish/GroupKeyQueue.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { GroupKey } from '../encryption/GroupKey'
+import { GroupKeyStore } from '../encryption/GroupKeyStore'
+
+export class GroupKeyQueue {
+
+    private currentGroupKey: GroupKey | undefined // current key id if any
```
Comment (harbu) [inline packages/client/src/publish/GroupKeyQueue.ts]: Is code comment up-to-date? Talking about `current key id` but here we have a `GroupKey`?

File: packages/client/src/publish/GroupKeyQueue.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    private queuedGroupKey: GroupKey | undefined // a group key queued to be rotated into use after the call to useGroupKey
+    private store: GroupKeyStore
```
Comment (harbu) [inline packages/client/src/publish/GroupKeyQueue.ts]:
```suggestion
    private readonly store: GroupKeyStore
```

File: packages/client/src/publish/Publisher.ts [hunk 3]
```diff
@@ -24,13 +27,21 @@ const parseTimestamp = (metadata?: MessageMetadata): number => {
...
 
     constructor(
         context: Context,
-        @inject(delay(() => PublishPipeline)) private pipeline: PublishPipeline
+        @inject(delay(() => PublishPipeline)) private pipeline: PublishPipeline,
+        groupKeyStoreFactory: GroupKeyStoreFactory
     ) {
         this.id = instanceId(this)
         this.debug = context.debug.extend(this.id)
+        this.getGroupKeyQueue = pMemoize(async (streamId: StreamID) => {
```
Comment (harbu) [inline packages/client/src/publish/Publisher.ts:39]: Why memoized?

Comment (teogeb) [reply packages/client/src/publish/Publisher.ts:39]: For each `streamId` we should keep one group key queue. This way we get the same instance for each call.

File: packages/client/src/publish/GroupKeyQueue.ts (hunk 1 above)
Comment (teogeb) [reply packages/client/src/publish/GroupKeyQueue.ts]: Fixed

File: packages/client/src/publish/GroupKeyQueue.ts (hunk 2 above)
Comment (teogeb) [reply packages/client/src/publish/GroupKeyQueue.ts]: Fixed

**Review by harbu — APPROVED** [review]

**Pull request merged**